frontend/app.py

```diff
@@ -0,0 +1,275 @@
+    decoded_data = data.decode("utf-8")
+
+    # Convert the decoded data into JSON format
+    data_json = json.loads(decoded_data)
+    
+    if 'error' in data_json:
+        raise ValueError(f"Error in getting access token: {data_json['error_description']}")
+    
+    if 'access_token' not in data_json:
+        raise KeyError(f"Access token not found in response. Response: {data_json}")
+    
+    access_token = data_json["access_token"]
+    return access_token
+
+def get_enphase_data(enphase_system_id: str, access_token: str) -> pd.DataFrame:
+    api_key = os.getenv('ENPHASE_API_KEY')
+    start_at = int((datetime.now() - timedelta(weeks=1)).timestamp())
+    granularity = "week"
+
+    conn = http.client.HTTPSConnection("api.enphaseenergy.com")
+    headers = {
+        "Authorization": f"Bearer {str(access_token)}",
+        "key": str(api_key)
+    }
+
+    url = f"/api/v4/systems/{enphase_system_id}/telemetry/production_micro?start_at={start_at}&granularity={granularity}"
+    conn.request("GET", url, headers=headers)
+    res = conn.getresponse()
+    data = res.read()
+    decoded_data = data.decode("utf-8")
+    data_json = json.loads(decoded_data)
+
+    return process_enphase_data(data_json, start_at)
+
+def enphase_authorization():
+    if 'access_token' not in st.session_state:
+        auth_url = get_enphase_auth_url()
+        st.write("Please visit the following URL to authorize the application:")
+        st.markdown(f"[Enphase Authorization URL]({auth_url})")
+        st.write("After authorization, you will be redirected to a URL. Please copy the entire URL and paste it below:")
+
+        redirect_url = st.text_input("Enter the redirect URL:")
+
+        if redirect_url:
+            if "?code=" not in redirect_url:
+                st.error("Invalid redirect URL. Please make sure you copied the entire URL.")
+                return None, None
+
+            auth_code = redirect_url.split("?code=")[1]
+
+            try:
+                access_token = get_enphase_access_token(auth_code)
+                st.session_state['access_token'] = access_token
+                return access_token, os.getenv('ENPHASE_SYSTEM_ID')
+            except Exception as e:
+                st.error(f"Error getting access token: {str(e)}")
+                return None, None
+    else:
+        return st.session_state['access_token'], os.getenv('ENPHASE_SYSTEM_ID')
+
+    return None, None
+
+def make_pv_data(site: PVSite, ts: pd.Timestamp, access_token: str = None, enphase_system_id: str = None) -> xr.Dataset:
+    live_generation_kw = None
+
+    if site.inverter_type == 'enphase' and access_token and enphase_system_id:
+        live_generation_kw = get_enphase_data(enphase_system_id, access_token)
+
+    da = process_pv_data(live_generation_kw, ts, site)
+    return da
+
+def predict_ocf(site: PVSite, model=None, ts: datetime | str = None, nwp_source: str = "icon", access_token: str = None, enphase_system_id: str = None):
```

Comment:
Could you just import this function? 

https://github.com/openclimatefix/Open-Source-Quartz-Solar-Forecast/blob/main/quartz_solar_forecast/forecast.py#L10
you might have to add enphase tokens to this function

Reply:
> Could you just import this function?
> 
> https://github.com/openclimatefix/Open-Source-Quartz-Solar-Forecast/blob/main/quartz_solar_forecast/forecast.py#L10 you might have to add enphase tokens to this function

that function indirectly uses the make_pv_data from data.py, that uses some of the CLI functions from enphase.py, since we want all the user inputs to be only on the web, i have modified these functions in this file